Make expansions stable for incr. comp.

cjgillot opened the pull request:
This PR aims to make expansions stable for incr. comp. by using the same architecture as definitions:
- the interned identifier `ExpnId` contains a `CrateNum` and a crate-local id;
- bidirectional maps `ExpnHash <-> ExpnId` are setup;
- incr. comp. on-disk cache saves and reconstructs expansions using their `ExpnHash`.

I tried to use as many `LocalExpnId` as I could in the resolver code, but I may have missed a few opportunities.

All this will allow to use an `ExpnId` as a query key, and to force this query without recomputing caller queries. For instance, this will be used to implement #85999.

r? @petrochenkov

rust-highfive commented:
Some changes occurred in src/tools/rustfmt.

cc @calebcartwright

cjgillot commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 48374baa140a0358612b4c1be2f7d8716620519d with merge 10d3de9d420a0b3963c1d7182283d29ff22412d5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"48374baa140a0358612b4c1be2f7d8716620519d","merge_sha":"10d3de9d420a0b3963c1d7182283d29ff22412d5"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2926691977)
Build commit: 10d3de9d420a0b3963c1d7182283d29ff22412d5 (`10d3de9d420a0b3963c1d7182283d29ff22412d5`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2926691977"},"merge_sha":"10d3de9d420a0b3963c1d7182283d29ff22412d5"} -->

rust-timer commented:
Queued 10d3de9d420a0b3963c1d7182283d29ff22412d5 with parent 9cdb2d3d59bad2843330535a41d0ebbb831fe57e, future [comparison URL](https://perf.rust-lang.org/compare.html?start=9cdb2d3d59bad2843330535a41d0ebbb831fe57e&end=10d3de9d420a0b3963c1d7182283d29ff22412d5).

rust-timer commented:
Finished benchmarking try commit (10d3de9d420a0b3963c1d7182283d29ff22412d5): [comparison url](https://perf.rust-lang.org/compare.html?start=9cdb2d3d59bad2843330535a41d0ebbb831fe57e&end=10d3de9d420a0b3963c1d7182283d29ff22412d5).

**Summary**: This change led to significant improvements 🎉 in compiler performance.
- Moderate improvement in [instruction counts]() (up to -1.7% on `full` builds of `deeply-nested-async-check`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

petrochenkov commented:
(I'll get to this next weekend, most likely.)

bors commented:
:umbrella: The latest upstream changes (presumably #86694) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #86982) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
Phew, I did a first pass over this PR just to understand what's going on, and it took a ~whole working day because the PR does a lot, it would be better to break it into parts if possible.

petrochenkov commented:
The fundamental difference between hashes used for definitions and expansions right now is that in case of definitions we a hashing something that is "resistant to small changes" (a def path), but in case of expansions we are just accumulating the whole `ExpnData` into the hash.
So if we, for example, move the location of a macro call slightly, its expansion hash will change as well.
But is doesn't really affects the infra (all the `ExpnHash <-> ExpnId` maps, etc), so we'll be able to change what exactly goes into the hash after landing this PR, that's good.

petrochenkov commented:
General comments:
- I'd prefer to see a newtype a la `DefIndex` for local indices instead of a raw `u32`s.
- Commits like https://github.com/rust-lang/rust/pull/86676/commits/e3abfd12f2e744ad4de888ab439f71a15c62e02e need comments, otherwise it's not clear why we are introducing duplication / copypaste for no apparent reason.
- It would be good to avoid back and forth like changing `ExpnId` to `LocalExpnId` in one commit and then changing it back in some later commit.
- Style nits: please use `impl Trait` in the first commit touching code in which it's usable; please don't use `Self` outside of generics, it almost always makes readability worse.

I doubt that it's possible to extract some "pure refactoring" parts from this PR, because the whole PR is kind of a gradual refactoring, but if it's possible, then it would be great.

cjgillot commented:
> Phew, I did a first pass over this PR just to understand what's going on, and it took a ~whole working day because the PR does a lot, it would be better to break it into parts if possible.

Sorry for the lack of explanations.

> The fundamental difference between hashes used for definitions and expansions right now is that in case of definitions we a hashing something that is "resistant to small changes" (a def path), but in case of expansions we are just accumulating the whole `ExpnData` into the hash.
> So if we, for example, move the location of a macro call slightly, its expansion hash will change as well.
> But it doesn't really affects the infra (all the `ExpnHash <-> ExpnId` maps, etc), so we'll be able to change what exactly goes into the hash after landing this PR, that's good.

I agree with this concern. I am trying to design an extension to #84373 which makes spans stable enough to resolve it.

> General comments:
> * I'd prefer to see a newtype a la `DefIndex` for local indices instead of a raw `u32`s.
> * Commits like e3abfd1 need comments, otherwise it's not clear why we are introducing duplication / copypaste for no apparent reason.
> * It would be good to avoid back and forth like changing  `ExpnId` to `LocalExpnId` in one commit and then changing it back in some later commit.
> * Style nits: please use `impl Trait` in the first commit touching code in which it's usable; please don't use `Self` outside of generics, it almost always makes readability worse.

I added a few commits and the current ones to avoid such back-and-forth. I could not find which functions you are referring to with `impl Trait`.

> I doubt that it's possible to extract some "pure refactoring" parts from this PR, because the whole PR is kind of a gradual refactoring, but if it's possible, then it would be great.

I extracted #87044 if it helps.

petrochenkov commented:
>I could not find which functions you are referring to with impl Trait.

`fn decode_expn_id` and all similar functions taking `impl Fn(Mut,Once)` callbacks as arguments.

petrochenkov commented:
Blocked on #87044.

bors commented:
:umbrella: The latest upstream changes (presumably #87042) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
>I extracted #87044 if it helps.

It certainly helped.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3061083701) [(plain)](https://github.com/rust-lang/rust/commit/b44c138329a213bd183fae6c24c3f143ab31613e/checks/3061083701/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_middle v0.0.0 (/checkout/compiler/rustc_middle)
    Checking rustc_ast_passes v0.0.0 (/checkout/compiler/rustc_ast_passes)
    Checking rustc_expand v0.0.0 (/checkout/compiler/rustc_expand)
    Checking rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error[E0609]: no field `cstore` on type `context::TyCtxt<'tcx>`
   --> compiler/rustc_middle/src/ty/query/on_disk_cache.rs:830:25
    |
830 |             decoder.tcx.cstore.expn_hash_to_expn_id(krate, index_guess, hash)
    |
    |
help: one of the expressions' fields has a field of the same name
    |
830 |             decoder.tcx.gcx.untracked_resolutions.cstore.expn_hash_to_expn_id(krate, index_guess, hash)

error: aborting due to previous error

For more information about this error, try `rustc --explain E0609`.
```

</details>

petrochenkov commented:
@cjgillot 
Do you plan to extract more parts, or this is ready for review?

cjgillot commented:
@petrochenkov I don't have any more parts to extract. This is ready for review.

cjgillot commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit fa79be2b76169f55a3d08ed368b849bf3bd50131 with merge 7ee5f35f1bc1b7b1cc7dd867921c1dfb6fe5f924...
<!-- homu: {"type":"TryBuildStarted","head_sha":"fa79be2b76169f55a3d08ed368b849bf3bd50131","merge_sha":"7ee5f35f1bc1b7b1cc7dd867921c1dfb6fe5f924"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3069302952)
Build commit: 7ee5f35f1bc1b7b1cc7dd867921c1dfb6fe5f924 (`7ee5f35f1bc1b7b1cc7dd867921c1dfb6fe5f924`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3069302952"},"merge_sha":"7ee5f35f1bc1b7b1cc7dd867921c1dfb6fe5f924"} -->

rust-timer commented:
Queued 7ee5f35f1bc1b7b1cc7dd867921c1dfb6fe5f924 with parent ee5ed4a88d6a869cdb152829ed697d6459650db3, future [comparison URL](https://perf.rust-lang.org/compare.html?start=ee5ed4a88d6a869cdb152829ed697d6459650db3&end=7ee5f35f1bc1b7b1cc7dd867921c1dfb6fe5f924).

petrochenkov commented on compiler/rustc_metadata/src/rmeta/encoder.rs:
```diff
@@ -31,7 +31,7 @@ use rustc_session::config::CrateType;
 use rustc_span::{self, ExternalSource, FileName, SourceFile, Span, SyntaxContext};
 use rustc_span::{
-    hygiene::{ExpnDataEncodeMode, HygieneEncodeContext, MacroKind},
+    hygiene::{ExpnIndex, HygieneEncodeContext, MacroKind},
```
Nit: could you avoid multi-line and nested imports?

petrochenkov commented on compiler/rustc_middle/src/ty/query/on_disk_cache.rs:
```diff
@@ -20,8 +20,7 @@ use rustc_serialize::{
 use rustc_span::hygiene::{
-    ExpnDataDecodeMode, ExpnDataEncodeMode, ExpnId, HygieneDecodeContext, HygieneEncodeContext,
-    SyntaxContext, SyntaxContextData,
+    ExpnId, HygieneDecodeContext, HygieneEncodeContext, SyntaxContext, SyntaxContextData,
```
Nit: could you avoid multi-line and nested imports?

rust-timer commented:
Finished benchmarking try commit (7ee5f35f1bc1b7b1cc7dd867921c1dfb6fe5f924): [comparison url](https://perf.rust-lang.org/compare.html?start=ee5ed4a88d6a869cdb152829ed697d6459650db3&end=7ee5f35f1bc1b7b1cc7dd867921c1dfb6fe5f924).

**Summary**: This benchmark run did not return any significant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up.  

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

petrochenkov commented on compiler/rustc_metadata/src/rmeta/encoder.rs:
```diff
@@ -176,12 +182,12 @@ impl<'a, 'tcx> Encodable<EncodeContext<'a, 'tcx>> for SyntaxContext {
-            s,
-        )
+        if self.krate == LOCAL_CRATE {
+            // We will only write details for local expansions.
```
That's true, but why?
What happens if we have a foreign expansion here and don't write details for it?

cjgillot replied:
In that case, the decoder will fetch the expansion from the other crate's metadata.

petrochenkov replied:
Aha, finally found that issue - https://github.com/rust-lang/rust/issues/73047 (and https://github.com/rust-lang/rust/issues/74731).

Sometimes it's preferable for the "other crate" to not exist in its `.rlib` form, proc macro crates and other "macro only" crates clearly "overstay their welcome" when they exist only for the spans to be decodable from them.
I also know that right now we simply drop some spans from proc macro (dependency?) crates so their rlibs are not required, but ideally they would be preserved.

So eventually we'll need some span metadata inlining scheme (which includes hygiene data).

petrochenkov replied:
Could you add a comment saying that here we are relying on the other crate's metadata existing, and maybe linking to these issues?

petrochenkov commented on compiler/rustc_span/src/hygiene.rs:
```diff
@@ -1020,11 +1137,18 @@ pub struct HygieneEncodeContext {
 
 impl HygieneEncodeContext {
+    /// Record the fact that we need to serialize the corresponding `ExpnData`.
+    pub fn schedule_for_encoding(&self, expn: ExpnId) {
```
```suggestion
    pub fn schedule_expn_data_for_encoding(&self, expn: ExpnId) {
```
I misinterpreted what this function does, the `ExpnId` itself is encoded immediately, it's `ExpnData` encoding which is delayed.

petrochenkov commented:
One remaining question about duplicate tables - https://github.com/rust-lang/rust/pull/86676#discussion_r669794333.
Although I now see that the local table's size is used for assigning fresh local IDs at least.

cjgillot commented:
> One remaining question about duplicate tables - [#86676 (comment)](https://github.com/rust-lang/rust/pull/86676#discussion_r669794333).
> Although I now see that the local table's size is used for assigning fresh local IDs at least.

I have two reasons for keeping separate foreign tables:
- they are expected to be sparse, where the local table is full;
- they should eventually be removed, to fetch the information directly from the metadata decoder like DefIds do.

If these are satisfactory, I will add them as comments.

petrochenkov commented:
r=me after squashing all the review stuff into main commits.

petrochenkov commented on compiler/rustc_hir/src/definitions.rs:
```diff
@@ -409,7 +409,7 @@ impl Definitions {
 
     pub fn expansion_that_defined(&self, id: LocalDefId) -> ExpnId {
-        self.expansions_that_defined.get(&id).copied().unwrap_or_else(ExpnId::root)
+        self.expansions_that_defined.get(&id).copied().unwrap_or(ExpnId::root())
```
Nit: unnecessary change.

cjgillot commented:
@bors r=petrochenkov

bors commented:
:pushpin: Commit b35ceeeec706e957ee46166e0a8b6d912ac215a3 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov b35ceeeec706e957ee46166e0a8b6d912ac215a3 -->
<!-- homu: {"type":"Approved","sha":"b35ceeeec706e957ee46166e0a8b6d912ac215a3","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit b35ceeeec706e957ee46166e0a8b6d912ac215a3 with merge 68511b574ffe019a5cb3e9fa92605f80d39167bc...
<!-- homu: {"type":"BuildStarted","head_sha":"b35ceeeec706e957ee46166e0a8b6d912ac215a3","merge_sha":"68511b574ffe019a5cb3e9fa92605f80d39167bc"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3094460796)
Approved by: petrochenkov
Pushing 68511b574ffe019a5cb3e9fa92605f80d39167bc to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3094460796"},"merge_sha":"68511b574ffe019a5cb3e9fa92605f80d39167bc"} -->
